stream_alert/apps/_apps/gsuite.py

```diff
@@ -151,11 +161,24 @@ def _gather_logs(self):
         # once during the first poll
         if not self._next_page_token:
             self._last_timestamp = activities[0]['id']['time']
-            LOGGER.debug('Caching last timestamp: %s', self._last_timestamp)
+            LOGGER.debug('[%s] Caching last timestamp: %s', self, self._last_timestamp)
+            # Store the event ids with the most recent timestamp to de-duplicate them next time
+            next_event_ids = [
+                activity['id']['uniqueQualifier']
+                for activity in activities
+                if activity['id']['time'] == self._last_timestamp
+            ]
+            self._context['next_event_ids'] = next_event_ids[:self._MAX_EVENT_IDS]
```

Comment:
or just something similar using a local property that then caches to the instance prop if you want to keep the warning message below

Reply:
The reason I needed to cache `_next_event_ids` and `_last_event_ids` is because the events are returned in reverse chronological order. The first few events contain the duplicates that we need to save, and the last events are the ones we want to de-duplicate. But in the case of a multiple data poll, there will be more than one invocation of `_gather_logs`, and the events at the end of the second poll will be the duplicates from the last polled time period, not necessarily the last lambda invocation. So we can't overwrite the cached event ids, and temporarily need to store both.

So, `_last_event_ids` are the event ids from the last poll, and `_next_event_ids` are the event ids for the next poll.

Reply:
Refactored to only store one copy in the context since _gather_logs is called multiple times per execution. The case where it gets deferred to the next lambda execution is a very small edge case, and worst case you get some duplicate events.